Utilities/DataCompression/include/DataCompression/DataDeflater.h

```
@@ -41,132 +64,130 @@ namespace AliceO2 {
-    uint16_t bitsWritten = 0;
+  template <typename ValueType, typename WriterT>
+  int writeRaw(ValueType value, uint16_t bitlength, WriterT writer) {
+    auto bitsToWrite = bitlength;
     if (bitlength > 8*sizeof(ValueType)) {
       // TODO: error policy
-      bitlength = 8*sizeof(ValueType);
+      throw std::runtime_error("bit length exceeds width of the data type");
     }
-    while (bitsWritten < bitlength) {
-      if (mCurrent == mEnd) {
-        //break; // depending on error policy
-        return -ENOSPC;
+    while (bitsToWrite > 0) {
+      if (mFilledBits == TargetBitWidth) {
+        mFilledBits=0;
+        writer(mCurrent);
+        mCurrent = 0;
       }
-      _TargetType& current = *mCurrent;
       // write at max what is left to be written
-      uint16_t writeNow = bitlength - bitsWritten;
-      ValueType mask = 1<<(writeNow); mask -= 1;
+      auto writeNow = bitsToWrite;
       // write one element of the target buffer at a time
-      if (writeNow > 8*sizeof(_TargetType)) writeNow = 8*sizeof(_TargetType);
+      if (writeNow > TargetBitWidth) writeNow = TargetBitWidth;
       // write the remaining space in the current element
-      uint16_t capacity=8*sizeof(_TargetType)-mBitPosition;
+      auto capacity = TargetBitWidth - mFilledBits;
       if (writeNow > capacity) writeNow = capacity;
-      ValueType activebits=(value&mask)>>(bitlength-bitsWritten-writeNow);
-      activebits<<(capacity-writeNow);
-      mBitPosition+=writeNow;
-      bitsWritten+=writeNow;
-      if (mBitPosition==8*sizeof(_TargetType)) {
-        mBitPosition=0;
-        mCurrent++;
-      } // pedantic check: should never exceed the taget type size
+      auto mask = (((ValueType)1 << writeNow) - 1) << (bitsToWrite - writeNow);
+      auto activebits = (value&mask) >> (bitsToWrite - writeNow);
+      mCurrent |= activebits << (capacity-writeNow);
+      mFilledBits += writeNow;
+      bitsToWrite -= writeNow;
+      assert(mFilledBits <= TargetBitWidth);
```

Is this a real assert or could it be triggered by bad input? If the latter I would throw an exception.